File: assets/js/socket/user_socket.js

import {Socket} from "phoenix"
let socket = new Socket("/user_socket", {params: {token: sessionStorage.userToken}})

socket.connect()
let user_id = sessionStorage.userSessionId
if (user_id) {
    let channel = socket.channel("user_session:"+user_id, {}) 
    channel.on("global_export:" + user_id + ":complete", ({ payload, fileName, extension }) => {
        var temp = "data:application/vnd.ms-excel;base64," + encodeURIComponent(payload);
        var download = document.createElement("a");
        download.href = temp;
        download.download = fileName + extension;
        document.body.appendChild(download);
        download.click();
        document.body.removeChild(download);
      });

    channel.join()

    // exports_channel.join()
}

export default socket

File: lib/loanmanagementsystem_web/channels/user_socket.ex
defmodule LoanmanagementsystemWeb.UserSocket do
  use Phoenix.Socket

  channel "user_session:*", LoanmanagementsystemWeb.UserChannel

  def join("user_session:" <> _user_id, _message, socket) do
    {:ok, socket}
  end

  def connect(%{"token" => token}, socket, _connect_info) do
    # max_age: 1209600 is equivalent to two weeks in seconds
    case Phoenix.Token.verify(socket, "user socket", token, max_age: 1_209_600) do
      {:ok, user_id} -> {:ok, assign(socket, :user, user_id)}
      {:error, _reason} -> :error
    end
  end

  @impl true
  def connect(_params, socket, _connect_info) do
    {:ok, socket}
  end


  @impl true
  def id(_socket), do: nil
end
